File: creational/prototype/molecular_simulation/typescript/molecular_simulation.ts
import { performance } from "perf_hooks"; // For more precise timing if needed, but console.time works

// Define the structure for parameters
export type ExperimentParameters = Record<string, any>;

// Represents the simulation object that acts as a prototype
export class MolecularSimulation {
  public moleculeName: string;
  public parameters: ExperimentParameters;
  private _precomputedStates: number[] = []; // Simulate large dataset
  private _expensiveSetupDone: boolean = false;

  // Simplified private constructor - only initializes basic properties.
  // State (_precomputedStates, _expensiveSetupDone) is set by create or clone.
  private constructor(moleculeName: string, parameters: ExperimentParameters) {
    this.moleculeName = moleculeName;
    this.parameters = structuredClone(parameters); // Deep clone params early
  }

  // Static method to create the initial prototype instance, ensuring setup runs.
  public static async create(
    moleculeName: string,
    baseParameters: ExperimentParameters
  ): Promise<MolecularSimulation> {
    console.log(`🧬 Initializing simulation for '${moleculeName}'...`);
    // Call simplified constructor
    const instance = new MolecularSimulation(moleculeName, baseParameters);

    // Perform expensive setup - create should be the only entry point needing this.
    // No need to check _expensiveSetupDone here if create is used correctly (once).
    const states = await instance._performExpensiveSetup();

    // Assign the computed states and mark setup as done AFTER it completes.
    instance._precomputedStates = states;
    instance._expensiveSetupDone = true;

    return instance;
  }

  // Simulates a time-consuming setup process and RETURNS the computed states.
  // Removed internal check for _expensiveSetupDone.
  private async _performExpensiveSetup(): Promise<number[]> {
    console.log(
      `⏳ Performing expensive precomputation for '${this.moleculeName}' (takes ~2 seconds)...`
    );

    // Simulate async delay (like file I/O or network request)
    await new Promise((resolve) => setTimeout(resolve, 2000));

    // Simulate generating large data
    const size = 1_000_000;
    const states = Array.from({ length: size }, () => Math.random() * 100);

    // Don't set flags here, just return the data
    console.log(
      `✅ Expensive setup complete for '${this.moleculeName}'. ${states.length} states computed.`
    );
    return states;
  }

  // The Prototype pattern's core method: clone
  public clone(): MolecularSimulation {
    console.log(`\n🔄 Cloning simulation for '${this.moleculeName}'...`);

    // Use the simplified constructor
    const clone = new MolecularSimulation(this.moleculeName, this.parameters); // Constructor deep copies params

    // Manually copy the state from the prototype AFTER construction
    clone._precomputedStates = this._precomputedStates; // Share the large data reference
    clone._expensiveSetupDone = true; // Explicitly mark setup as done for the clone

    console.log(`    Cloned simulation created. Setup Skipped.`);
    return clone;
  }

  // Modify a specific parameter for this simulation instance
  public setParameter(key: string, value: any): void {
    console.log(
      `    Setting parameter '${key}' = ${JSON.stringify(value)} for '${
        this.moleculeName
      }' simulation`
    );
    this.parameters[key] = value;
  }

  // Runs the simulation using its current parameters and precomputed data
  public run(): void {
    console.log(
      `\n🔬 Running simulation for '${
        this.moleculeName
      }' with parameters: ${JSON.stringify(this.parameters)}`
    );
    if (!this._precomputedStates || this._precomputedStates.length === 0) {
      console.error("   ❌ Error: Precomputed states not available!");
      return;
    }

    // Simulate work using the data and parameters
    const temp = this.parameters["temperature"] ?? 298.15; // Kelvin
    const pressure = this.parameters["pressure"] ?? 1.0; // atm
    const duration = this.parameters["duration"] ?? 100; // picoseconds

    // Example calculation (placeholder)
    const resultMetric =
      ((this._precomputedStates.slice(0, 1000).reduce((a, b) => a + b, 0) *
        (temp / 273.15)) /
        pressure) *
      (duration / 10);
    console.log(
      `   Simulation complete. Result metric: ${resultMetric.toFixed(2)}`
    );
    console.log(
      `   (Used ${this._precomputedStates.length} precomputed states)`
    );
  }

  // Getter for testing
  public getPrecomputedStatesLength(): number {
    return this._precomputedStates.length;
  }
}

File: creational/prototype/molecular_simulation/python/molecular_simulation.py
import time
import copy
import random
from typing import List, Dict, Any

# Define a simple parameter structure (can be more complex)
ExperimentParameters = Dict[str, Any]

class MolecularSimulation:
    """Represents a potentially expensive simulation setup (Concrete Prototype)."""

    def __init__(self, molecule_name: str, base_parameters: ExperimentParameters):
        """Initializes the simulation, including the expensive setup."""
        self.molecule_name = molecule_name
        self.parameters = copy.deepcopy(base_parameters) # Ensure base params are copied
        self._precomputed_states: List[float] = []

        print(f"🧬 Initializing simulation for '{self.molecule_name}'...")
        self._perform_expensive_setup()

    def _perform_expensive_setup(self):
        """Simulates a time-consuming setup process."""
        print(f"⏳ Performing expensive precomputation for '{self.molecule_name}' (takes ~2 seconds)...")
        # Simulate delay
        time.sleep(2)
        # Simulate generating large data
        self._precomputed_states = [random.random() * 100 for _ in range(1_000_000)] # Simulate large data
        print(f"✅ Expensive setup complete for '{self.molecule_name}'. {len(self._precomputed_states)} states computed.")

    def clone(self) -> 'MolecularSimulation':
        """Creates a clone, sharing expensive data and deep copying parameters."""
        print(f"\n🔄 Cloning simulation for '{self.molecule_name}'...")

        # 1. Create a new instance without calling __init__ (to skip expensive setup)
        # Using __new__ is one way, another is a dedicated internal method if needed.
        new_simulation = self.__class__.__new__(self.__class__)

        # 2. Copy essential attributes
        new_simulation.molecule_name = self.molecule_name
        # Deep copy parameters to ensure independence
        new_simulation.parameters = copy.deepcopy(self.parameters)

        # 3. Share the reference to the expensive precomputed data
        new_simulation._precomputed_states = self._precomputed_states

        # Note: Python doesn't explicitly need the '_expensiveSetupDone' flag here
        # because skipping __init__ via __new__ prevents the setup call.

        print(f"    Cloned simulation created. Setup Skipped.")
        return new_simulation

    def set_parameter(self, key: str, value: Any):
        """Modify a specific parameter for this simulation instance."""
        print(f"    Setting parameter '{key}' = {value} for '{self.molecule_name}' simulation")
        self.parameters[key] = value

    def run(self):
        """Runs the simulation using its current parameters and precomputed data."""
        print(f"\n🔬 Running simulation for '{self.molecule_name}' with parameters: {self.parameters}")
        if not self._precomputed_states:
            print("   ❌ Error: Precomputed states not available!")
            return

        # Simulate work using the data and parameters
        temp = self.parameters.get('temperature', 298.15) # Kelvin
        pressure = self.parameters.get('pressure', 1.0) # atm
        duration = self.parameters.get('duration', 100) # picoseconds

        # Example calculation using precomputed data and params
        # (This is just a placeholder for actual simulation logic)
        result_metric = sum(self._precomputed_states[:1000]) * (temp / 273.15) / pressure * (duration / 10)
        print(f"   Simulation complete. Result metric: {result_metric:.2f}")
        print(f"   (Used {len(self._precomputed_states)} precomputed states)")
